smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/serde/readwrite/WritingClosureUtils.kt

```diff
@@ -0,0 +1,84 @@
+import software.amazon.smithy.swift.codegen.integration.serde.json.TimestampUtils
+import software.amazon.smithy.swift.codegen.integration.serde.xml.NodeInfoUtils
+import software.amazon.smithy.swift.codegen.model.getTrait
+import software.amazon.smithy.swift.codegen.model.hasTrait
+
+class WritingClosureUtils(
+    val ctx: ProtocolGenerator.GenerationContext,
+    val writer: SwiftWriter
+) {
+
+    private val nodeInfoUtils = NodeInfoUtils(ctx, writer)
+
+    fun writingClosure(member: MemberShape): String {
+        val target = ctx.model.expectShape(member.target)
+        val memberTimestampFormatTrait = member.getTrait<TimestampFormatTrait>()
+        return writingClosure(target, memberTimestampFormatTrait)
+    }
+
+    fun writingClosure(shape: Shape): String {
+        return writingClosure(shape, null)
+    }
+
+    private fun writingClosure(shape: Shape, memberTimestampFormatTrait: TimestampFormatTrait? = null): String {
+        if (ctx.service.hasTrait<AwsJson1_0Trait>() || ctx.service.hasTrait<AwsJson1_1Trait>() || ctx.service.hasTrait<RestJson1Trait>()) {
+            return "JSONReadWrite.writingClosure()"
+        } else if (ctx.service.hasTrait<AwsQueryTrait>() || ctx.service.hasTrait<Ec2QueryTrait>()) {
+            return "FormURLReadWrite.writingClosure()"
+        }
+        when (shape) {
+            is MapShape -> {
+                val keyNodeInfo = nodeInfoUtils.nodeInfo(shape.key)
+                val valueNodeInfo = nodeInfoUtils.nodeInfo(shape.value)
+                val valueWriter = writingClosure(shape.value)
+                val isFlattened = shape.hasTrait<XmlFlattenedTrait>()
+                return writer.format(
+                    "SmithyXML.mapWritingClosure(valueWritingClosure: \$L, keyNodeInfo: \$L, valueNodeInfo: \$L, isFlattened: \$L)",
+                    valueWriter,
+                    keyNodeInfo,
+                    valueNodeInfo,
+                    isFlattened
+                )
+            }
+
+            is ListShape -> {
+                val memberNodeInfo = nodeInfoUtils.nodeInfo(shape.member)
+                val memberWriter = writingClosure(shape.member)
+                val isFlattened = shape.hasTrait<XmlFlattenedTrait>()
+                return writer.format(
+                    "SmithyXML.listWritingClosure(memberWritingClosure: \$L, memberNodeInfo: \$L, isFlattened: \$L)",
+                    memberWriter,
+                    memberNodeInfo,
+                    isFlattened
+                )
+            }
+
+            is TimestampShape -> {
+                return writer.format(
+                    "SmithyXML.timestampWritingClosure(format: \$L)",
+                    TimestampUtils.timestampFormat(memberTimestampFormatTrait, shape)
+                )
+            }
+
+            else -> {
+                return writer.format("\$N.writingClosure(_:to:)", ctx.symbolProvider.toSymbol(shape))
+            }
+        }
+    }
+}
```

Comment:
Kotlin best practice is to lift return statements out such that you do `return when { ... }` and then the individual `return` in each block can be removed completely as the last statement in a code block is always returned.

you can also replace the first `if ... else if ... else` with a when statement to make it more idiomatic.

In all, I suggest refactoring this whole function to the following:

```
private fun writingClosure(shape: Shape, memberTimestampFormatTrait: TimestampFormatTrait? = null): String {
    when {
        ctx.service.hasTrait<AwsJson1_0Trait>() || 
        ctx.service.hasTrait<AwsJson1_1Trait>() || 
        ctx.service.hasTrait<RestJson1Trait>() -> return "JSONReadWrite.writingClosure()"

        ctx.service.hasTrait<AwsQueryTrait>() || 
        ctx.service.hasTrait<Ec2QueryTrait>() -> return "FormURLReadWrite.writingClosure()"
    }

    return when (shape) {
        is MapShape -> {
            val keyNodeInfo = nodeInfoUtils.nodeInfo(shape.key)
            val valueNodeInfo = nodeInfoUtils.nodeInfo(shape.value)
            val valueWriter = writingClosure(shape.value)
            val isFlattened = shape.hasTrait<XmlFlattenedTrait>()
            writer.format(
                "SmithyXML.mapWritingClosure(valueWritingClosure: \$L, keyNodeInfo: \$L, valueNodeInfo: \$L, isFlattened: \$L)",
                valueWriter, keyNodeInfo, valueNodeInfo, isFlattened
            )
        }

        is ListShape -> {
            val memberNodeInfo = nodeInfoUtils.nodeInfo(shape.member)
            val memberWriter = writingClosure(shape.member)
            val isFlattened = shape.hasTrait<XmlFlattenedTrait>()
            writer.format(
                "SmithyXML.listWritingClosure(memberWritingClosure: \$L, memberNodeInfo: \$L, isFlattened: \$L)",
                memberWriter, memberNodeInfo, isFlattened
            )
        }

        is TimestampShape -> writer.format(
            "SmithyXML.timestampWritingClosure(format: \$L)",
            TimestampUtils.timestampFormat(memberTimestampFormatTrait, shape)
        )

        else -> writer.format("\$N.writingClosure(_:to:)", ctx.symbolProvider.toSymbol(shape))
    }
}
```

Reply:
Fixed as suggested